A decorator for Metrics that works without SimpleTraits (#52)
JuliaNLSolvers/Manifolds.jl: docs/src/manifolds/metric.md

# Metric manifold

A Riemannian manifold always consists of a [topological manifold](https://en.wikipedia.org/wiki/Topological_manifold) together with a smoothly varying metric $g$.

> However, often there is a implicitly assumed (daefault) metric, like the usual

Reviewer: ```suggestion
However, often there is an implicitly assumed (default) metric, like the usual
```

inner product on [`Euclidean`](@ref) space. This decorator takes this into
account. Hence it is not necessary to acts as a (as transparent as possible)

> decorator, see [`is_decorator_manifold`](@ref) for details.

Reviewer: I don't really understand this sentence ("Hence it (...) for details"). Could you rephrase it?

> This manifold servers two purposes: To implement different metrics (i.g. in closed

Reviewer: ```suggestion
This manifold decorator serves two purposes: to implement different metrics (e.g. in closed
```

form) for one [`Manifold`](@ref) and to provide a way to compute geodesics

> on manifolds, where this [`Metric`](@ref) does not yield closed formula. 

Reviewer: I think eventually all other functions would also have a default implementation based on the metric, right? Like `exp`, `log`, parallel transport etc.

Author: Well I hope so, yes, for now there is only exp implemented based on the metric tensor. I have not yet read how to do the remaining ones numerically – but also since my goal is more to implement different metrics

Let's first look at the provided types.

> ## types

Reviewer: ```suggestion
## Types
```

```@autodocs
Modules = [Manifolds, ManifoldsBase]
Pages = ["Metric.jl"]
Order = [:type]
```

## Implement Different Metrics on the same Manifold

In order to distinguish different metrics on one manifold, one can introduce
two [`Metric`](@ref)s and use this type to dispatch on the metric, see 
[`SymmetricPositiveDefinite`](@ref). To avoid overhead, one [`Metric`](@ref)
can then be marked as being the default, i.e. the one that is used, when no

> [`MetricManifold`](@ref) is used. This avoids reimplementation of a first

Reviewer: ```suggestion
[`MetricManifold`](@ref) is used. This avoids reimplementation of the first
```

> existing metric, access for the default metric functions (without an [`MetricManifold`](@ref)

Reviewer: ```suggestion
existing metric, access to the default metric functions (without an [`MetricManifold`](@ref)
```

Perhaps also "default metric functions" should be replaced by a more precise phrase, maybe "metric-dependent functions of the undecorated manifold"?

> overhead) as well as the transparent fallback of the corresponding [`MetricManifold`](@ref)

Reviewer: Just to clarify, `MetricManifold` used in this way should not cause any runtime overhead.

Author: Yes, it should not cause overhead. I'll add that.

> with default metric to the (classical) implementations.

Reviewer: Maybe "undecorated" would be clearer than "classical"?

> ## Implementaion of Metrics

Reviewer: ```suggestion
## Implementation of Metrics
```

# Metric manifold

A Riemannian manifold always consists of a [topological manifold](https://en.wikipedia.org/wiki/Topological_manifold) together with a smoothly varying metric $g$.

However, often there is an implicitly assumed (default) metric, like the usual
inner product on [`Euclidean`](@ref) space. This decorator takes this into
account. It is not necessary to use this decorator, if you imlement just one (or

> the first) metric. If you later introduce a second, the old (first) one can be

Author: I avoided the second "one" by rephrasing the sentence.

> used with or without a metricx. The decorator acts transparent in that sense;

Reviewer: ```suggestion
used with or without a metric. The decorator acts transparent in that sense;
```
